### ISSUE TITLE
`maybe_mark_dynamic` causes max recursion error when used with compile during tensordict consolidation

### ISSUE THREAD

**Issue by vmoens** [pytorch/pytorch#138729]
## Context

We would like to consolidate tensordict made of NJTs in a single storage.

This should contain all the values, offsets and lengths of all the leaves in the TD. The operation looks like this
```python
swaps = copy(tensors)
for i, t in enumerate(tensors):
    swaps[i] = tensor.view(torch.uint8).view(-1)
torch.stack(swaps, storage)
out = storage.split(tensor_lengths)
out = [t.view(t_src.dtype).view(t_src.shape) for t, t_src in zip(out, tensors)]
```

This operation can be hugely CPU-overhead bound and when compiled we can get a nice 2x speed-up. It should work with dynamic shapes (for torchrec users for instance you may know how many jagged tensors you'll get ahead of time but not how big they are going to be).

## Issue

Currently, doing that when the number of NJTs exceeds 64 or so causes a max recursion depth error.

The line responsible for this is this one:
https://github.com/pytorch/pytorch/blob/cd9c6e9408dd79175712223895eed36dbdc84f84/torch/nested/_internal/nested_tensor.py#L145C9-L145C41

I created a minimal reproducible example for this that only requires PT to be installed:
https://gist.github.com/vmoens/c15e6a6862ad849e30856034cbf33f1a

If you comment out the line 8:
```python
        torch._dynamo.maybe_mark_dynamic(values, 0)
```
then the code runs fine (although compile time can be exceedingly long). 

With this line uncommented, you should get:

```
torch._dynamo.exc.TorchRuntimeError: Failed running call_method view(*(FakeTensor(..., size=(4*s141,), dtype=torch.uint8), torch.float32), **{}):
maximum recursion depth exceeded while calling a Python object

from user code:
   File "/Users/vmoens/Library/Application Support/JetBrains/PyCharm2023.3/scratches/scratch_8.py", line 57, in consolidate
    result = [
  File "/Users/vmoens/Library/Application Support/JetBrains/PyCharm2023.3/scratches/scratch_8.py", line 58, in <listcomp>
    view_old_as_new(v, oldv)
  File "/Users/vmoens/Library/Application Support/JetBrains/PyCharm2023.3/scratches/scratch_8.py", line 52, in view_old_as_new
    v = v.view(oldv.dtype)
```

cc @ezyang @chauhang @penguinwu @eellison @bobrenjc93 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @rec @zou3519 @bdhirsh @yf225 @jbschlosser @anijain2305

**Comment by jbschlosser** [comment]
Looking into this a bit now.. the `torch._dynamo.maybe_mark_dynamic(values, 0)` line causes a new symbol to be allocated for each tensor shape (`s0` through `s255` for the repro). Without this line, the same `s0` symbol is allocated for all tensors, since they have the same shape. The formula for the storage offset of each tensor involves these symbols and becomes increasingly unwieldy. For example:
```
4*s0 + 4*s1 + 4*s10 + 4*s100 + 4*s101 + 4*s102 + 4*s103 + 4*s104 + 4*s105 + 4*s106 + 4*s107 + 4*s108 + 4*s109 + 4*s11 + 4*s110 + 4*s111 + 4*s112 + 4*s113 + 4*s114 + 4*s115 + 4*s116 + 4*s117 + 4*s118 + 4*s119 + 4*s12 + 4*s120 + 4*s121 + 4*s122 + 4*s123 + 4*s124 + 4*s125 + 4*s126 + 4*s127 + 4*s128 + 4*s129 + 4*s13 + 4*s130 + 4*s131 + 4*s132 + 4*s133 + 4*s134 + 4*s135 + 4*s136 + 4*s137 + 4*s138 + 4*s139 + 4*s14 + 4*s140 + 4*s15 + 4*s16 + 4*s17 + 4*s18 + 4*s19 + 4*s2 + 4*s20 + 4*s21 + 4*s22 + 4*s23 + 4*s24 + 4*s25 + 4*s26 + 4*s27 + 4*s28 + 4*s29 + 4*s3 + 4*s30 + 4*s31 + 4*s32 + 4*s33 + 4*s34 + 4*s35 + 4*s36 + 4*s37 + 4*s38 + 4*s39 + 4*s4 + 4*s40 + 4*s41 + 4*s42 + 4*s43 + 4*s44 + 4*s45 + 4*s46 + 4*s47 + 4*s48 + 4*s49 + 4*s5 + 4*s50 + 4*s51 + 4*s52 + 4*s53 + 4*s54 + 4*s55 + 4*s56 + 4*s57 + 4*s58 + 4*s59 + 4*s6 + 4*s60 + 4*s61 + 4*s62 + 4*s63 + 4*s64 + 4*s65 + 4*s66 + 4*s67 + 4*s68 + 4*s69 + 4*s7 + 4*s70 + 4*s71 + 4*s72 + 4*s73 + 4*s74 + 4*s75 + 4*s76 + 4*s77 + 4*s78 + 4*s79 + 4*s8 + 4*s80 + 4*s81 + 4*s82 + 4*s83 + 4*s84 + 4*s85 + 4*s86 + 4*s87 + 4*s88 + 4*s89 + 4*s9 + 4*s90 + 4*s91 + 4*s92 + 4*s93 + 4*s94 + 4*s95 + 4*s96 + 4*s97 + 4*s98 + 4*s99
```

In the repro, what finally breaks is `view()` attempting an int_floordiv of this by 4. Looking into possible fixes / workarounds now.

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/138660 just landed, see if it helps

**Comment by ezyang** [comment]
@vmoens says it doesn't help, which is understandable; you need to actually call torch.sym_sum to sum these symbols together.

**Comment by vmoens** [comment]
Amazing, it compiled! (took 6 mins for 128 small NJTs)

**Comment by ezyang** [comment]
@laithsakka probably still some asymptotic badness here, I am guessing

**Comment by vmoens** [comment]
Yeah so to give you an idea, compile time for 
8 NJTs is 4.35
16 NJTs is 10 secs
32 NJTs is 57 sec
64 NJTs is 329 sec

So doubling the number of tensors more than quadruples the compile time.

Roughly the same pattern occurs with non-NJTs (where we don't have the recursion issue) though compile times are smaller

**Comment by masnesral** [comment]
[issue scrubbing] @laithsakka you were kinda the last one mentioned here. Do you know if this is still an issue? Seems like the original problem was resolved, but there's some asymptotic behavior? (I'm not sure how to repro myself to see if it's still a problem).

**Comment by laithsakka** [comment]
@masnesral 
1) Max recursion error is solved by https://github.com/pytorch/pytorch/issues/141215
2)The other perf issues is related to  https://github.com/pytorch/pytorch/issues/140835. and blocked on it, 
this did not seem to be hi pri when discussing with the team recently. @bobrenjc93 @bdhirsh 

TLDR; 2 is probably solvable, but we do not have bandwidth for it now and the priority does not seem very high

**Comment by laithsakka** [comment]
i think we can close this since we have an issue for the views-re-application
